[API] Changes to mutation API

akhenry opened the pull request:
Changes how object mutation works behind the scenes in order to keep objects in sync automatically when their model changes.

* The way that objects are mutated and observed has not changed, `openmct.objects.mutate` and `openmct.objects.observe` should still be used in the same way that they were before.
* Behind the scenes, domain objects that are mutable are wrapped in a new [`MutableDomainObject`](https://github.com/nasa/openmct/pull/3483/files?diff=split&w=1#diff-7ef3a5358f05632acb4ff900098f346817a36e1759a4dc36972629b36afeddd2) that exposes mutator and observer functions that allow objects to be mutated in such a way that all instances can be kept in sync.
* It is now possible to retrieve `MutableDomainObject`s from the API, instead of regular domain objects. These are automatically updated when mutation occurs on any instance of the object, replacing the need for "`*`" listeners. Note that __the view API now provides objects in this form by default__. Therefore, you do not need to do anything differently in views, the domain objects will just magically keep themselves up to date.
* If for some reason you need to retrieve an object manually via `openmct.objects.get` (you should ask why you need to do this) and you want it to magically keep itself in sync, there is a new API function named [`openmct.objects.getMutable(identifier)`](https://github.com/nasa/openmct/pull/3483/files?diff=split&w=1#diff-7b3a993b89a077474a9a4cf7fca79c18131e8e222951d164d2848ab147afbd24R177). Note that if you do this **you will be responsible for the object's lifecycle**. It relies on listeners which must be destroyed when the object is no longer needed, otherwise memory leaks will occur. You can destroy a `MutableDomainObject` and its (internal) listeners by calling [`openmct.objects.destroyMutable(mutableDomainObject)`](https://github.com/nasa/openmct/pull/3483/files?diff=split&w=1#diff-7b3a993b89a077474a9a4cf7fca79c18131e8e222951d164d2848ab147afbd24R193). Any listeners created by calls to `openmct.objects.observe`  need to be cleaned up separately.
* If the composition of a `MutableDomainObject` is retrieved using the Composition API, [all children will be returned as `MutableDomainObjects` automatically](https://github.com/nasa/openmct/pull/3483/files?diff=split&w=1#diff-a8e739b28dcaffbcb94a984957ada70c7c386eafabb2a3537c5d632dc495f9fdR198). Their lifecycle will be [managed automatically, and is tied to the lifecycle of the parent](https://github.com/nasa/openmct/pull/3483/files?diff=split&w=1#diff-a8e739b28dcaffbcb94a984957ada70c7c386eafabb2a3537c5d632dc495f9fdR67).
* Any `MutableDomainObject` provided by the Open MCT framework itself (eg. provided to view providers by the View API, or from the composition API) will have its lifecycle managed by Open MCT, you don't need to worry destroying it.
* [Some](https://github.com/nasa/openmct/pull/3483/files?diff=split&w=1#diff-c81b5971bfc1744860f666d448e5be2aac414b8e60d5f3e6e75f5fdc8b8fdfd7L91) [instances](https://github.com/nasa/openmct/pull/3483/files?diff=split&w=1#diff-075bd86f4339e7ff7550d10d7e0d660a135bc9558c4d1ad315b533da9336c3d8L208) of "`*`" observers have been now been replaced. Due to time constraints I was unable to replace all of them, so I will file followup issues to address them.
* There were (and still are) a large number of views where we are tracking multiple instances of the same domain object (usually named things like `observedDomainObject`, `internalDomainObject`, etc.). I have removed these wherever I have changed view code, but there are still a lot of examples of this in our codebase. We should avoid having multiple different instances of the same domain object in the same view, it's very confusing to developers. This was I think partly necessitated by the need to use "`*`" observers, so hopefully we won't need to do it going forward.

__TL;DR__
* "`*`" observers are no longer needed anywhere. In views, the domain object provided to the view provider will automatically be kept in sync, so if the domain object is a reactive Vue property (eg. it's passed in via `props`) then Vue reactivity will take care of updating the view automatically.
* If for some reason you need to manually retrieve a MutableDomainObject (ie. an automatically synchronized version of a domain object) you can use `openmct.objects.getMutable(identifier)`. However, you will be responsible for destroying it.

Note: I updated the Object API to be an ES6 module. This was necessitated by a weird issue in Webpack with using a mixture of require and ES6 imports, that I could not resolve using the usual `.default` fix. This resulted the indentation of the whole file shifting, which means every line shows up as changed in the diff unless you use the `Hide whitespace changes` option in the File Changes tab.


To do:
* [x] Add an overview of the changes
* [x] File followup issues
    * Fix remaining "`*`" listeners https://github.com/nasa/openmct/issues/3553
    * Bug with Plot labels - https://github.com/nasa/openmct/issues/3554

akhenry commented on src/plugins/LADTable/components/LADRow.vue:
```diff
@@ -26,7 +26,7 @@
     @contextmenu.prevent="showContextMenu"
 >
-    <td class="js-first-data">{{ name }}</td>
+    <td class="js-first-data">{{ domainObject.name }}</td>
```
`domainObject` provided by view API is now automatically kept in sync, so I can just use Vue reactivity to update the name when it changes. No need for any event listeners at all.

akhenry commented on src/plugins/displayLayout/components/DisplayLayout.vue:
```diff
@@ -152,29 +152,26 @@ export default {
     computed: {
         gridSize() {
-            return this.internalDomainObject.configuration.layoutGrid;
+            return this.domainObject.configuration.layoutGrid;
```
No longer necessary to track two different instances of an object (eg. `internalDomainObject` or `observedDomainObject` vs . just `domainObject`).

davetsay commented on src/api/composition/CompositionCollection.js:
```diff
@@ -189,6 +195,15 @@ define([
+            if (this.returnMutables && this.publicAPI.objects.supportsMutation(child)) {
+                let keyString = this.publicAPI.objects.makeKeyString(child.identifier);
+
+                if (this.publicAPI.objects.supportsMutation(child)) {
```
Wouldn't this always be true given the outer if block?

akhenry replied:
You appear to be correct. Let me take a look and de-crappify accordingly 😄

davetsay commented on src/api/objects/MutableDomainObject.js:
```diff
@@ -0,0 +1,126 @@
+/*****************************************************************************
+ * Open MCT, Copyright (c) 2014-2019, United States Government
```
Daaaaamn. You started working on this in 2019? Mad props

akhenry replied:
Yeah, good catch 😄  It's long overdue. Started as a branch of topic-core-refactor, kept getting out of date, and would take a couple of hours each time to get everything working again.

The issue it fixes is actually even older than this and dates back to [an epic debate in a PR from 2016](https://github.com/nasa/openmct/pull/1131#issuecomment-240855549)...

davetsay commented on src/ui/router/Browse.js:
```diff
@@ -36,10 +38,19 @@ define([
 
-            openmct.layout.$refs.browseObject.show(object, viewProvider.key, true, currentObjectPath);
-            openmct.layout.$refs.browseBar.domainObject = object;
+            openmct.layout.$refs.browseObject.show(mutable || object, viewProvider.key, true, currentObjectPath);
+            openmct.layout.$refs.browseBar.domainObject = mutable || object;
```
I know objects truthiness is pretty absolute but shouldn't we be explicitly checking if its defined?

davetsay commented on src/api/objects/ObjectAPI.js:
```diff
@@ -20,337 +20,387 @@
+        //Mutate original object
+        MutableDomainObject.mutateObject(domainObject, path, value);
+
+        //Mutate temptary mutable object, in the process informing any other mutable instances
```
spelling error

akhenry replied:
Actually it's only for mutating tempting or "temptary" (as I call them) domain objects.

shefalijoshi commented on src/api/objects/ObjectAPI.js:
```diff
@@ -20,337 +20,387 @@
+ObjectAPI.prototype.supportsMutation = function (object) {
+    // Checking for mutability is a bit broken right now. This is an 80% solution,
+    // but does not work in some cases.
+    const type = this.typeRegistry.get(object.type);
```
For which cases does checking for mutability not work?

akhenry replied:
eg. using folders to group things in a telemetry dictionary. The folder type is create-able (it appears in the create menu), but it shouldn't be mutable in the case of a telemetry dictionary. Our API doesn't allow us to distinguish between these two things right now, and we have been using "creatability" as a proxy for this for a long time. We need a better solution, but it's out of scope for this change. Maybe object providers should provide a hint about whether the objects they provide are mutable? 
eg.
```JavaScript
isMutable(objectIdentifier) {
    return false; // Or implement some more complicated logic here. Maybe the object provider returns a mix of mutable and non-mutable objects? I dunno.
}
```

At least if it's behind an API function for now it will make changing the implementation a little easier later on.

jvigliotta replied:
Would this fix that: https://github.com/nasa/openmct/issues/3612 ? @akhenry

akhenry replied:
@jvigliotta Yes, it will address the example case.

shefalijoshi commented on src/api/objects/ObjectAPI.js:
```diff
@@ -20,337 +20,387 @@
+        let mutableDomainObject = this._toMutable(domainObject);
+
+        //Mutate original object
+        MutableDomainObject.mutateObject(domainObject, path, value);
```
Do we need to call`MutableDomainObject.mutateObject(domainObject, path, value);` before we call `mutableDomainObject.$set(path, value)`? 
It appears that mutableDomainObject.$set already does what the static method does here: https://github.com/nasa/openmct/pull/3483/files?file-filters%5B%5D=.js&hide-deleted-files=true#diff-7ef3a5358f05632acb4ff900098f346817a36e1759a4dc36972629b36afeddd2R72

akhenry replied:
The problem is that creating a `MutableDomainObject` actually creates a copy of the non-mutable `domainObject` that is passed in. So calling `$set` on it will mutate the copy, and emit an event so that any other `MutableDomainObject` instances of the same object get updated, but it will not actually mutate the original `domainObject`. 

"Why does it create a copy?" you might very reasonably ask. The aim was to stick to the principle that domain object are just plain old JavaScript objects, without any functions or special properties added by the framework. This was one of the foundational principles of the refactor away from the original API.

It achieves this via inheritance, which requires a new object to be created that inherits the special functions and properties of the `MutableDomainObject` without them existing on the object itself.

shefalijoshi commented on src/api/objects/ObjectAPI.js:
```diff
@@ -20,337 +20,387 @@
-                this.mutate(domainObject, 'persisted', persistedTime);
-                result = provider.update(domainObject);
-            }
+            return object;
```
Returning a non-mutable object here seems misleading. We should throw an error if the object does not support mutation.

akhenry replied:
You're right, I'll change it.

akhenry replied:
Oh wait, now I remember 😄 There is a problem with that because there is no way of knowing whether an object is mutable or not based solely on its identifier, you have to fetch it first, so it's a catch-22. 

Let me come back to this. Maybe `isPersistable` is a workable proxy for mutability that doesn't require a fully resolved object?

akhenry replied:
Fixed it using `isPersistable`. `getMutable` will only return mutables, and will error otherwise.

shefalijoshi commented on src/api/objects/ObjectAPI.js:
```diff
@@ -20,337 +20,387 @@
+ */
+ObjectAPI.prototype.observe = function (domainObject, path, callback) {
+    if (domainObject.isMutable) {
+        return domainObject.$observe(path, callback);
```
Why don't we return `() => domainObject.$destroy();` here like in the `else` block?

akhenry replied:
[MutableDomainObject.$observe returns a destroy function](https://github.com/nasa/openmct/blob/mutation-update/src/api/objects/MutableDomainObject.js#L69).

shefalijoshi commented on src/api/objects/MutableDomainObject.js:
```diff
@@ -0,0 +1,126 @@
+
+        mutable.$observe('$_synchronize_model', (updatedObject) => {
+            let clone = JSON.parse(JSON.stringify(updatedObject));
+            let deleted = _.difference(Object.keys(updatedObject), Object.keys(updatedObject));
```
Is there a typo here? Should _.difference be checking `updatedObject` and `mutable`?

akhenry replied:
Yes, almost certainly, I'll fix it. Thanks

akhenry replied:
Fixed.

shefalijoshi reviewed (dismissed):
Finished halfway, but going to submit this for comments/changes anyway.

akhenry commented:
Well, that was a Homeric epic. 

I believe I have addressed all of the review comments, merged from master, and fixed the infinity bugs that came out of it.

akhenry commented:
Oops, missed one feedback comment. Fixed now.

shefalijoshi commented:
@akhenry 
I think the changes look good, but I'm not sure what you meant by this comment. Which view API are you talking about here?
`Note that the view API now provides objects in this form by default. Therefore, you do not need to do anything differently in views, the domain objects will just magically keep themselves up to date.`

akhenry commented:
> @akhenry
> I think the changes look good, but I'm not sure what you meant by this comment. Which view API are you talking about here?
> `Note that the view API now provides objects in this form by default. Therefore, you do not need to do anything differently in views, the domain objects will just magically keep themselves up to date.`

The Open MCT View API. Views are now provided with a MutableDomainObject to their `view` function instead of a regular `DomainObject`. Because the lifecycle of a view is well defined Open MCT itself can manage the destruction of the `MutableDomainObject` that it provides to views, so that the views themselves don't need to worry about it.

shefalijoshi commented:
@akhenry please resolve conflict so we can merge this.

akhenry commented:
@shefalijoshi Done!

davetsay reviewed (approved):
LGTM

## Reviewer Checklist
| | |
| --- | :---: |
| Changes appear to address issue? | Y |
| Appropriate unit tests included? | Y |
| Code style and in-line documentation are appropriate? | Y |
| Commit messages meet standards? | Y |
| Has associated issue been labelled `unverified`? | Y |

davetsay commented:
@akhenry , merged, but can you provide an issue with testing instructions?
